OpenTTD/OpenTTD
# Feature: store table headers for each chunk in savegame

## Motivation / Problem

So there is a bit of a story here.

First off, the main reason I started this change, is because I wanted to move around some settings. But the order of settings that are stored in the savegame cannot be changed. That is really annoying, as there isn't really way around that. Sure, we can change the order, do a savegame version bump .. but that also leaves us with a huge list of the original order to make sure old games load. I wanted this fixed.

Next, we quickly figured out this was an universal issue with our savegame format. It allows for no modification of order of anything .. often leaving the code a bit weird and odd, because of what is stored on disk. This tightly-coupled information has made many mistakes in the past.

Lastly, for a while now we have somewhat the ambition to move to SQLite as savegame format. But getting there from where we were, is far from easy.
I am the type of person that doesn't like to go from A to Z immediately, but I do like that the changes we make are at least on the route towards Z. This PR not only proves SQLite should be possible, but it also prepares our SaveLoad code enough that SQLite would be possible.

Those three things combined gave me the motivation (hihi, same word, different context) to write this PR.

## Description

To decouple what fields are on disk and what we are expected, I basically made the savegames self-descriptive. With that I mean there is enough information in the savegame to know how to read the savegame, without having to know the savegame version or any internals about how OpenTTD loads the game .. well, other than the format, ofc.

The idea is simple: all chunks (with a few exceptions, see below) are now `CH_TABLE`, indicating they are a table. A table has a header and a list of entries (shocker). In other words, the old `CH_ARRAY` prefixed with a header.

The header is very easy: a list of (type, key) pairs, telling in what order the keys are stored, and what type the field is. The list ends with a type=0 entry.
If there is a struct inside that list, the header of that struct is directly followed by the header you are reading (so, it is depth-first). This gives you full view of the entries in the chunk.

The type-field is a hard-coded list, and can have 1 flag: is this is a list or not. If it is a list, when reading the entry the field will be start with a gamma telling you how many items to expect.

So how does OpenTTD react on headers that are different than what he expects?

- If the order of fields is different, it will adjust the order to match what is in the savegame. So you are free to move around settings, members, etc.
- If a field type is changed, loading fails. The developer should have bumped the version and made the conversion code as we already do.
- If a field exists in the savegame but you don't know it, it will be skipped reading.
- If a field doesn't exist in the savegame but you expect it, it will be skipped reading.

These last two are rather important: if the adding or removing of a field influences the gameplay, you should also bump the savegame version. Often this will mean: only removing doesn't need a bump, adding most likely does. But this is up to the developer.

To keep backwards compatibility, in `src/saveload/compat` are now tables with the original order of fields. As this wasn't stored in the savegame, this table is used.

Finally, to top off this PR, I created https://github.com/TrueBrain/OpenTTD-savegame-reader . This is a very simply Python application that only knows the savegame format. It doesn't know anything about any chunks. And still it can read savegames made with this PR fine. The reader is meant as verification that the client indeed creates savegames that can easily be read by external tools.
As bonus, it allows us to export savegames as JSON.
And because I love weird shit, I also made it for the web: https://truebrain.github.io/OpenTTD-savegame-reader/
:D

## Testing

I did some extensive testing with this:

I have a test-suite of ~500 savegames, ranging from TTDp savegame up to 1.11.2. A total of over 100 different savegame versions.
For every savegame I did:
- Save the old savegame, call it "main".
- Load the old savegame (call it "original"), run it for 256 ticks. Save the game, export as JSON.
- Load the "main", run it for 256 ticks. Save the game, export as JSON.
- Compare the JSON files. There should be no difference, in theory.

And except some minor totally unrelated bugs, this is in fact the case. That shows that at the very least everything that was loaded is saved and loaded correctly again.

Lastly, I did a lot of testing by saving a game with this PR, and then removing fields from the code. That all seems to work out fine .. I can even remove complete structs, and it will do the right thing.

## Limitations

- MAP-chunks (with the exception of `MAPS`) are still a `CH_RIFF`. I could not think of a sane and useful way to convert them to tables, so I left them alone.
- The compat tables only mention the order of the fields, not the type the field was original in. This could be an addition to this PR, but honestly, not sure that really adds value.
- `AIPL` and `GSTR` contain "junk" at the end of each entry. This is the blob of data saved/loaded by AI/GSes in Squirrel. In theory this could be put in a table too, but the free format it has makes it a bit difficult. To not make this PR an endless deep hole, I kinda scoped that out of this PR for now. I might come back to this decision.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)
